Pgp flow

yhcote opened the pull request:
Better look and feel when using sign/verify and internal cleanup of code in signing and openpgp stack

yhcote committed:
Look and fix key store file mode if necessary.

Signed-off-by: Yannick Cote <yhcote@gmail.com>

yhcote committed:
Use the askQuestion() func to cleanup code.

Signed-off-by: Yannick Cote <yhcote@gmail.com>

yhcote committed:
Various code cleanup.

Signed-off-by: Yannick Cote <yhcote@gmail.com>

yhcote committed:
Aim at last SIF commit.

Signed-off-by: Yannick Cote <yhcote@gmail.com>

yhcote committed:
Export AskQuestion() routine

Signed-off-by: Yannick Cote <yhcote@gmail.com>

yhcote committed:
Better the Sign workflow and general cleanup.

Signed-off-by: Yannick Cote <yhcote@gmail.com>

yhcote committed:
Better verification flow and some cleanup.

Signed-off-by: Yannick Cote <yhcote@gmail.com>

yhcote committed:
Start working on multi signature verification.

Signed-off-by: Yannick Cote <yhcote@gmail.com>

yhcote committed:
Merge branch 'master' of https://github.com/singularityware/singularity into pgp-flow

yhcote committed:
Fix typo

Signed-off-by: Yannick Cote <yhcote@gmail.com>

yhcote committed:
Improve messaging.

Signed-off-by: Yannick Cote <yhcote@gmail.com>

tri-adam commented on src/pkg/signing/signing.go:
```diff
@@ -22,23 +22,13 @@ import (
 
 const (
-	keyserverURI = "https://keys.sylabs.io:11371"
+	defaultKeysServer = "https://keys.sylabs.io:11371"
```
Can you remove `:11371`? We can use https://keys.sylabs.io now, port 443 will work from everywhere (Starbucks with paranoid firewall etc.)

tri-adam commented on src/pkg/sypgp/sypgp.go:
```diff
@@ -107,15 +108,28 @@ func printSignatures(entity *openpgp.Entity) error {
-		sylog.Errorf("could lookup user's real home folder %s\n", err)
-		return ""
+		sylog.Warningf("could not lookup user's real home folder %s\n", err)
+		sylog.Warningf("using current directory for %s\n", filepath.Join(".singularity", "sypgp"))
```
Nitpick: logs shouldn't end with `\n`.

bauerm97 replied:
Huh, so I just realized that sylog strips `\n` from the ending of any log messages, and then makes sure there is exactly one `\n` at the end. So there's a lot of code not doing this properly all over the rest of the codebase!

tri-adam replied:
Yeah, it's an idiom change from v2->v3, I'm sure these will be pervasive for a while. If you like, we could start linting for that 😉

yhcote replied:
Yes and this could have been a leftover from the previous sylog call. sypgp was an early sylog adopter ;-) This is the right way to do it though and I believed I lobbied (with support) for that to happen also.

A general code cleanup is a good idea @bauerm97 !

tri-adam reviewed (changes requested):
Two tiny requests

yhcote committed:
params to client and doc cleanup.

Signed-off-by: Yannick Cote <yhcote@gmail.com>

yhcote committed:
Merge branch 'master' of https://github.com/singularityware/singularity into pgp-flow

yhcote committed:
Minor output improvement.

Signed-off-by: Yannick Cote <yhcote@gmail.com>

tri-adam commented on src/cmd/singularity/cli/keys_pull.go:
```diff
@@ -18,7 +18,7 @@ import (
 func init() {
 	KeysPullCmd.Flags().SetInterspersed(false)
-	KeysPullCmd.Flags().StringVarP(&url, "url", "u", "", "overwrite the default remote url")
+	KeysPullCmd.Flags().StringVarP(&keyServerURL, "url", "u", "", "specify the key server URL (default: https://keys.sylabs.io)")
```
`StringVarP` fourth argument is the default, let's use that. And rather than duplicate the URL, let's use the variable. This:

```
KeysPullCmd.Flags().StringVarP(&keyServerURL, "url", "u", defaultKeysServer, "specify the key server URL")
```

Will give you the desired result:

```
  -u, --url string   specify the key server URL (default
                     "https://keys.sylabs.io")
```

yhcote replied:
Didn't even realize from looking at other commands help that the default output came from printing the string var with Cobra! Alright.

tri-adam commented on src/cmd/singularity/cli/keys_push.go:
```diff
@@ -18,7 +18,7 @@ import (
 func init() {
 	KeysPushCmd.Flags().SetInterspersed(false)
-	KeysPushCmd.Flags().StringVarP(&url, "url", "u", "", "overwrite the default remote url")
+	KeysPushCmd.Flags().StringVarP(&keyServerURL, "url", "u", "", "specify the key server URL (default: https://keys.sylabs.io)")
```
Same here:

```
KeysPushCmd.Flags().StringVarP(&keyServerURL, "url", "u", defaultKeysServer, "specify the key server URL")
```

tri-adam commented on src/cmd/singularity/cli/keys_search.go:
```diff
@@ -18,7 +18,7 @@ import (
 func init() {
 	KeysSearchCmd.Flags().SetInterspersed(false)
-	KeysSearchCmd.Flags().StringVarP(&url, "url", "u", "", "overwrite the default remote url")
+	KeysSearchCmd.Flags().StringVarP(&keyServerURL, "url", "u", "", "overwrite the default remote url")
```
Missing the URL. Suggest same approach as the others here.

tri-adam commented on src/cmd/singularity/cli/sign.go:
```diff
@@ -17,26 +17,43 @@ import (
 
 func init() {
 	SignCmd.Flags().SetInterspersed(false)
+	SignCmd.Flags().StringVarP(&keyServerURL, "url", "u", "", "specify the key server URL (default: https://keys.sylabs.io)")
```
Suggest same approach as the others here.

tri-adam commented on src/cmd/singularity/cli/sign.go:
```diff
@@ -17,26 +17,43 @@ import (
+
+func doSignCmd(cpath, url string) error {
+	if url == "" {
+		// lookup key management server URL from singularity.conf
```
Overriding with the `singularity.conf` value here might be confusing to the user. Let's say I run `singularity sign -h`, and I see:

```
  -u, --url string   specify the key server URL (default:
                     https://keys.sylabs.io)
```

So now I think, OK, that default is not what I want. I want it to go to `https://otherkeys.sylabs.io`, so I do `singularity sign -u https://otherkeys.sylabs.io`. Now, unbeknownst to me, `singularity.conf` has `https://yetanotherkeys.sylabs.io` in it. I run that, and it sends my key to `https://yetanotherkeys.sylabs.io`.

I think the behaviour I would expect in the above scenario, is that the value in `singularity.conf` would affect the _default_, but then if I explicitly set `-u`, it overrides that default. Does that make sense?

yhcote replied:
Yeah, that logic is not done. Only a comment as a reminder that maybe we wanna setup a list of key servers in singularity.conf. I will remove that entire block for now since default + the usage of -u is functional until we decide on a format to store a list of keys in singularity.conf or a file similar.

tri-adam commented on src/cmd/singularity/cli/verify.go:
```diff
@@ -17,19 +17,20 @@ import (
 
 func init() {
 	VerifyCmd.Flags().SetInterspersed(false)
+	VerifyCmd.Flags().StringVarP(&keyServerURL, "url", "u", "", "specify the key server URL (default: https://keys.sylabs.io)")
```
Suggest same approach here, use the built in default capability of `StringVarP`.

tri-adam commented on src/cmd/singularity/cli/verify.go:
```diff
@@ -40,3 +41,18 @@ var VerifyCmd = &cobra.Command{
+		// lookup key management server URL from singularity.conf
+
+		// else use default builtin
+		url = defaultKeysServer
```
Same user confusion comment here. Suggest putting this logic in `init()`.

yhcote replied:
This is removed completely see my comment above.

tri-adam commented on src/pkg/signing/signing.go:
```diff
@@ -78,25 +64,49 @@ func sifAddSignature(fingerprint [20]byte, fimg *sif.FileImage, signature []byte
+		} else {
+			return fmt.Errorf("cannot sign without installed keys")
+		}
+		resp, err = sypgp.AskQuestion("Upload public key %X to key server? [Y/n] ", entity.PrimaryKey.Fingerprint)
```
It would be nice to see the URL that will be used in this message.

yhcote replied:
Great now it's wrapping to the next output line ;-) Done.

tri-adam commented on src/cmd/singularity/cli/keys_search.go:
```diff
@@ -18,7 +18,7 @@ import (
 func init() {
 	KeysSearchCmd.Flags().SetInterspersed(false)
-	KeysSearchCmd.Flags().StringVarP(&url, "url", "u", "", "overwrite the default remote url")
+	KeysSearchCmd.Flags().StringVarP(&keyServerURL, "url", "u", "", "overwrite the default remote url")
```
Should the `search` also read key server default from config file like `sign` and `verify`?

yhcote replied:
This is removed completely see my comment above.

tri-adam commented on src/cmd/singularity/cli/keys_push.go:
```diff
@@ -18,7 +18,7 @@ import (
 func init() {
 	KeysPushCmd.Flags().SetInterspersed(false)
-	KeysPushCmd.Flags().StringVarP(&url, "url", "u", "", "overwrite the default remote url")
+	KeysPushCmd.Flags().StringVarP(&keyServerURL, "url", "u", "", "specify the key server URL (default: https://keys.sylabs.io)")
```
Should the `push` also read key server default from config file like `sign` and `verify`?

yhcote replied:
This is removed completely see my comment above.

tri-adam commented on src/cmd/singularity/cli/keys_pull.go:
```diff
@@ -18,7 +18,7 @@ import (
 func init() {
 	KeysPullCmd.Flags().SetInterspersed(false)
-	KeysPullCmd.Flags().StringVarP(&url, "url", "u", "", "overwrite the default remote url")
+	KeysPullCmd.Flags().StringVarP(&keyServerURL, "url", "u", "", "specify the key server URL (default: https://keys.sylabs.io)")
```
Should `pull` also read key server default from config file like `sign` and `verify`?

yhcote replied:
This is removed completely see my comment above.

yhcote committed:
Fix the use of cobra params.

This commit also removes block marked with comments about
singularity.conf. This logic will be redefined in a subsequent commit.
It doesn't belong here.

Signed-off-by: Yannick Cote <yhcote@gmail.com>

tri-adam commented on src/cmd/singularity/cli/sign.go:
```diff
@@ -17,26 +17,36 @@ import (
 var SignCmd = &cobra.Command{
 	DisableFlagsInUseLine: true,
-	Args:   cobra.ExactArgs(1),
+	Args:   cobra.RangeArgs(1, 2),
```
I think this should be `cobra.ExactArgs(1)`, as `sign` only takes one positional argument.

tri-adam commented on src/cmd/singularity/cli/verify.go:
```diff
@@ -17,19 +17,20 @@ import (
 var VerifyCmd = &cobra.Command{
 	DisableFlagsInUseLine: true,
-	Args:   cobra.ExactArgs(1),
+	Args:   cobra.RangeArgs(1, 2),
```
I think this should be `cobra.ExactArgs(1)`, as `verify` only takes one positional argument.

tri-adam commented on src/cmd/singularity/cli/keys_search.go:
```diff
@@ -27,7 +27,7 @@ var KeysSearchCmd = &cobra.Command{
 	DisableFlagsInUseLine: true,
```
Github won't let me comment on the correct line, but think this command should be `cobra.ExactArgs(1)` also.

tri-adam commented on src/cmd/singularity/cli/keys_push.go:
```diff
@@ -27,7 +27,7 @@ var KeysPushCmd = &cobra.Command{
 	DisableFlagsInUseLine: true,
```
Same here, should be `cobra.ExactArgs(1)`.

tri-adam commented on src/cmd/singularity/cli/keys_pull.go:
```diff
@@ -27,7 +27,7 @@ var KeysPullCmd = &cobra.Command{
 	DisableFlagsInUseLine: true,
```
Same here, should be `cobra.ExactArgs(1)`.

tri-adam reviewed (changes requested):
OK, last round I think, just fixing up `cobra.Command.Args`. Sorry I didn't catch these last round!

yhcote committed:
Fix the magical Cobra # of argument usage.

Signed-off-by: Yannick Cote <yhcote@gmail.com>

yhcote committed:
Seems like we don't have any positional args here.

Signed-off-by: Yannick Cote <yhcote@gmail.com>
